lib/timeline-window.js

```diff
@@ -0,0 +1,392 @@
+
+    this._eventCount = 0;
+    this._windowLimit = opts.windowLimit || 1000;
+}
+
+/**
+ * Initialise the window to point at a given event, or the live timeline
+ *
+ * @param {string} [initialEventId]   If given, the window will contain the
+ *    given event
+ * @param {number} [initialWindowSize = 20]   Size of the initial window
+ *
+ * @return {module:client.Promise}
+ */
+TimelineWindow.prototype.load = function(initialEventId, initialWindowSize) {
+    var self = this;
+    initialWindowSize = initialWindowSize || 20;
+
+    var prom;
+    if (initialEventId) {
+        debuglog("TimelineWindow: initialising for event " + initialEventId);
+        prom = this._client.getEventTimeline(this._room, initialEventId)
+            .then(function(tl) {
+                // make sure that our window includes the event
+                for (var i = 0; i < tl.getEvents().length; i++) {
+                    if (tl.getEvents()[i].getId() == initialEventId) {
+                        return [tl, i];
+                    }
+                }
+                throw new Error("getEventTimeline result didn't include requested event");
+            });
+    } else {
+        debuglog("TimelineWindow: initialising with live timeline");
+
+        // start with the most recent events
+        var tl = this._room.getLiveTimeline();
+        prom = q([tl, tl.getEvents().length]);
+    }
+
+    prom = prom.then(function(v) {
+        var tl = v[0], eventIndex = v[1];
+
+        var endIndex = Math.min(tl.getEvents().length,
+                                eventIndex + initialWindowSize / 2);
+        var startIndex = Math.max(0, endIndex - initialWindowSize);
+        self._start = new TimelineIndex(tl, startIndex - tl.getBaseIndex());
+        self._end = new TimelineIndex(tl, endIndex - tl.getBaseIndex());
+        self._eventCount = endIndex - startIndex;
+    });
+
+    return prom;
+};
+
+/**
+ * Check if this window can be extended
+ *
+ * <p>This returns true if we either have more events, or if we have a
+ * pagination token which means we can paginate in that direction. It does not
+ * necessarily mean that there are more events available in that direction at
+ * this time.
+ *
+ * @param {string} direction   EventTimeline.BACKWARDS to check if we can
+ *   paginate backwards; EventTimeline.FORWARDS to check if we can go forwards
+ *
+ * @return {boolean} true if we can paginate in the given direction
+ */
+TimelineWindow.prototype.canPaginate = function(direction) {
+    var tl;
+    if (direction == EventTimeline.BACKWARDS) {
+        tl = this._start;
+        if (tl.index > tl.minIndex()) { return true; }
+    } else if (direction == EventTimeline.FORWARDS) {
+        tl = this._end;
+        if (tl.index < tl.maxIndex()) { return true; }
+    } else {
+        throw new Error("Invalid direction '" + direction + "'");
+    }
+
+    return tl.timeline.getNeighbouringTimeline(direction) ||
+        tl.timeline.getPaginationToken(direction);
```

Comment:
This violates your return type. Please wrap with `Boolean()`. `||` doesn't coerce to a boolean for you, but returns the first truthy object (which is useful for setting default values).